pkg/ingress/controller/controller.go

```diff
@@ -0,0 +1,548 @@
+	switch {
+	case errors.IsNotFound(err):
+		log.WithFields(log.Fields{"key": key}).Info("ingress has been deleted, will delete octavia resources")
+		err = c.deleteIngress(namespace, name)
+		if err != nil {
+			utilruntime.HandleError(fmt.Errorf("failed to delete openstack resources for ingress %s: %v", key, err))
+		}
+	case err != nil:
+		return fmt.Errorf("error fetching object with key %s from store: %v", key, err)
+	default:
+		if !IsValid(ing) {
+			log.WithFields(log.Fields{"key": key}).Info("ignore ingress")
+			return nil
+		}
+
+		log.WithFields(log.Fields{"key": key, "ingress": ing}).Info("ingress created or updated, will create or update octavia resources")
+
+		err = c.ensureIngress(ing)
+		if err != nil {
+			utilruntime.HandleError(fmt.Errorf("failed to create openstack resources for ingress %s: %v", key, err))
+		}
+	}
+
+	return nil
+}
+
+func (c *Controller) deleteIngress(namespace, name string) error {
+	lbName := fmt.Sprintf("k8s-%s-%s-lb", namespace, name)
+	loadbalancer, err := c.osClient.GetLoadbalancerByName(lbName)
+	if err != nil {
+		if err != openstack.ErrNotFound {
+			return fmt.Errorf("error getting loadbalancer %s: %v", name, err)
+		}
+		log.WithFields(log.Fields{"lbName": lbName, "ingressName": name, "ingressNamespace": namespace}).Info("loadbalancer for ingress deleted")
+		return nil
+	}
+
+	if err = c.osClient.DeleteFloatingIP(loadbalancer.VipPortID); err != nil {
+		return err
+	}
+
+	return c.osClient.DeleteLoadbalancer(loadbalancer.ID)
+}
+
+func (c *Controller) ensureIngress(ing *ext_v1beta1.Ingress) error {
+	var lbName = getResourceName(ing, "lb")
+	lb, err := c.osClient.EnsureLoadBalancer(lbName, c.config.Octavia.SubnetID)
+	if err != nil {
+		return err
+	}
+
+	var listenerName = getResourceName(ing, "listener")
+	listener, err := c.osClient.EnsureListener(listenerName, lb.ID)
+	if err != nil {
+		return err
+	}
+
+	// get nodes information
+	nodeObjs, err := c.nodeLister.ListWithPredicate(getNodeConditionPredicate())
+	if err != nil {
+		return err
+	}
+
+	// Add default pool for the listener if 'backend' is defined
+	defaultPoolName := listenerName + "-pool"
+	if ing.Spec.Backend != nil {
+		serviceName := fmt.Sprintf("%s/%s", ing.ObjectMeta.Namespace, ing.Spec.Backend.ServiceName)
+		nodePort, err := c.getServiceNodePort(serviceName, ing.Spec.Backend.ServicePort.IntValue())
+		if err != nil {
+			return err
+		}
+
+		if _, err = c.osClient.EnsurePoolMembers(false, defaultPoolName, "", listener.ID, &nodePort, nodeObjs); err != nil {
+			return err
+		}
+	} else {
+		// Delete default pool and its members
+		if _, err = c.osClient.EnsurePoolMembers(true, defaultPoolName, lb.ID, listener.ID, nil, nil); err != nil {
+			return err
+		}
+	}
+
+	// Delete all existing policies
+	existingPolicies, err := c.osClient.GetL7policies(listener.ID)
+	if err != nil {
+		return err
+	}
+	for _, p := range existingPolicies {
+		c.osClient.DeleteL7policy(p.ID, lb.ID)
```

Comment:
wonder why we can't run a delta? Removing all the policies before adding will take additional time

Reply:
My first version implementation was using delta handling, but in that way we need to query existing policies and rules, compare with new policies/rules each time the ingress is updated, the logic is complex and also involve many api requests. Considering updating ingress is not such a frequent operation, it's acceptable for now IMHO. However, I agree with you that we definitely need to improve this when we really receive complaints from our customers.